src/runtime_src/hip/core/memory_pool.cpp

```diff
@@ -0,0 +1,407 @@
+      throw std::runtime_error("mem poolsize is too big.");
+    }
+    else if (m_pool_size == m_max_total_size)
+    {
+      m_auto_extend = false;
+    }
+
+    m_list.emplace(m_list.end(), std::make_shared<memory_pool_node>(m_device, m_pool_size, m_last_id++));
+  }
+
+  void
+  memory_pool::get_attribute(hipMemPoolAttr attr, void* value)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
+
+    switch (attr)
+    {
+    case hipMemPoolReuseFollowEventDependencies:
+      *reinterpret_cast<int*>(value) = m_reuse_follow_event_dependencies;
+      break;
+
+    case hipMemPoolReuseAllowOpportunistic:
+      *reinterpret_cast<int*>(value) = m_reuse_allow_opportunistic;
+      break;
+
+    case hipMemPoolReuseAllowInternalDependencies:
+      *reinterpret_cast<int*>(value) = m_reuse_allow_internal_dependencies;
+      break;
+
+    case hipMemPoolAttrReleaseThreshold:
+      *reinterpret_cast<uint64_t*>(value) = m_release_threshold;
+      break;
+
+    case hipMemPoolAttrReservedMemCurrent:
+      *reinterpret_cast<uint64_t*>(value) = m_reserved_mem_current;
+      break;
+
+    case hipMemPoolAttrReservedMemHigh:
+      *reinterpret_cast<uint64_t*>(value) = m_reserved_mem_high;
+      break;
+    
+    case hipMemPoolAttrUsedMemCurrent:
+      *reinterpret_cast<uint64_t*>(value) = m_used_mem_current;
+      break;
+
+    case hipMemPoolAttrUsedMemHigh:
+      *reinterpret_cast<uint64_t*>(value) = m_used_mem_high;
+      break;
+    };
+  }
+
+  void
+  memory_pool::set_attribute(hipMemPoolAttr attr, void* value)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
+
+    switch (attr)
+    {
+    case hipMemPoolReuseFollowEventDependencies:
+      m_reuse_follow_event_dependencies = *reinterpret_cast<int*>(value);
+      break;
+
+    case hipMemPoolReuseAllowOpportunistic:
+      m_reuse_allow_opportunistic = *reinterpret_cast<int*>(value);
+      break;
+
+    case hipMemPoolReuseAllowInternalDependencies:
+      m_reuse_allow_internal_dependencies = *reinterpret_cast<int*>(value);
+      break;
+
+    case hipMemPoolAttrReleaseThreshold:
+      m_release_threshold = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrReservedMemCurrent:
+      m_reserved_mem_current = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrReservedMemHigh:
+      m_reserved_mem_high = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrUsedMemCurrent:
+      m_used_mem_current = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrUsedMemHigh:
+      m_used_mem_high = *reinterpret_cast<uint64_t*>(value);
+      break;
+    };
+  }
+
+  bool
+  memory_pool::extend_memory_list(size_t size)
+  {
+    m_list.insert(m_list.begin(), std::make_shared<memory_pool_node>(m_device, size, m_last_id++));
+    return true;
+  }
+
+  void memory_pool::malloc(void**ptr, size_t size)
+  {
+    *ptr = alloc(size);
+  }
+
+  void*
+  memory_pool::alloc(size_t size)
```

Comment:
This function is much too long.  Anything longer than 15 lines require significant commenting.   Prefer refactoring so that code reads well even without comments.

Reply:
This function is shortened and more comment has been added.